### PR TITLE
CUMULUS-2856: Allow removing execution fields

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-2856: Update PUT /executions endpoint to allow removing fields](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2856)

## Changes

* Update API/Message write logic to handle nulls as deletion in execution PUT/message write logic
* Update execution PUT endpoint to no longer respect message write constraints and update all values passed in

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: packages/es-client/indexer.js [hunk 1]
```diff
@@ -317,7 +350,7 @@ async function upsertGranule({
 }, writeConstraints = true) {
   Object.keys(updates).forEach((key) => {
     if (updates[key] === null && granuleInvalidNullFields.includes(key)) {
-      throw new ValidationError(`Attempted DynamoDb write with invalid key ${key} set to null.  Please remove or change this field and retry`);
```
Comment (Jkovarik) [inline packages/es-client/indexer.js:320]: Oops!    Good catch.

File: packages/es-client/tests/test-execution-upsert.js [hunk 2]
```diff
@@ -44,15 +45,17 @@ test.serial('upsertExecution writes new "running" execution', async (t) => {
     index: esIndex,
   });
 
+  delete testRecord.randomKey;
   const record = await esExecutionsClient.get(testRecord.arn);
   t.like(record, testRecord);
 });
 
-test.serial('upsertExecution writes new "completed" execution', async (t) => {
+test.serial('upsertExecution writes new "completed" execution with null fields omitted', async (t) => {
```
Comment (Jkovarik) [inline packages/es-client/tests/test-execution-upsert.js:53]: We need a test that shows nullification happens on overwrite as well as new writes.

File: packages/types/api/executions.d.ts [hunk 3]
```diff
@@ -7,9 +7,10 @@ export type ExecutionProcessingTimes = {
 
 export interface ExecutionRecord {
```
Comment (Jkovarik) [inline packages/types/api/executions.d.ts:8]: Should these types have a similar name, e.g. ApiExecutionRecord and ApiRecord for consistency with granules, pdrs, etc?

File: packages/types/api/executions.d.ts [hunk 4]
```diff
@@ -21,6 +22,25 @@ export interface ExecutionRecord {
   tasks?: object,
   timestamp?: number,
   type?: string,
-  updatedAt: number,
   cumulusVersion?: string,
 }
+
+export interface ApiExecution {
```
Comment (Jkovarik) [inline packages/types/api/executions.d.ts:28]: Nit: Alpha sort

File: CHANGELOG.md [hunk 5]
```diff
@@ -81,6 +81,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   - **CUMULUS-2792**
     - Remove API Granule model and all related test code
     - Remove granule-csv endpoint
+  - **CUMULUS-2856**
+    - Update API/Message write logic to handle nulls as deletion in execution PUT/message write logic
+    - Update execution PUT endpoint to no longer respect message write constraints and update all values passed in
```
Comment (Jkovarik) [inline CHANGELOG.md:86]: This might be a breaking change

**Review by Jkovarik — CHANGES_REQUESTED** [review]
This looks good @jennyhliu , I just have a couple of suggestions/questions re: tests.   Let me know if you want to pair/talk through any of it!

**Comment by jennyhliu** [comment]
@Jkovarik I have addressed your comments in [05e4e6b](https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9) [a8c31f75](https://github.com/nasa/cumulus/pull/3186/commits/a8c31f75c73982d49a634a5342571ed3627ee901)

File: CHANGELOG.md (hunk 5 above)
Comment (Jkovarik) [reply CHANGELOG.md]: Resolved in https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9

File: packages/types/api/executions.d.ts (hunk 3 above)
Comment (Jkovarik) [reply packages/types/api/executions.d.ts]: Resolved https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9

File: packages/db/src/translate/executions.ts [hunk 6]
```diff
@@ -67,12 +71,36 @@ export const translatePostgresExecutionToApiExecution = async (
   return <ExecutionRecord>removeNilProperties(translatedRecord);
 };
 
+const returnNullOrUndefinedOrDate = (
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/db/src/translate/executions.ts:74]: @jennyhliu  I think this comment was missed.   It's not *super* important, but for maintainability we could probably put this in .... lib?     Mostly wanted to get your thoughts on the comment.

File: packages/api/lib/writeRecords/write-execution.js [hunk 7]
```diff
@@ -85,19 +85,21 @@ const buildExecutionRecord = ({
...
-  });
+  };
+  return omitBy(record, isUndefined);
 };
 
 const writeExecutionToES = async (params) => {
   const {
     apiRecord,
     esClient = await Search.es(),
+    writeConstraints,
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/writeRecords/write-execution.js]: Resolved https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9

File: packages/api/tests/endpoints/test-executions.js [hunk 8]
```diff
@@ -1424,6 +1426,146 @@ test.serial('PUT /executions updates the record as expected in PostgreSQL/Elasti
...
+    tasks: { fakeTask: 'fake' },
+  });
+
+  const updatedExecution = fakeExecutionFactoryV2({
+    ...execution,
+    status: 'running',
+    asyncOperationId: null,
+    collectionId: null,
+    cumulusVersion: null,
+    duration: 180.5,
```
Comment (Jkovarik) [inline packages/api/tests/endpoints/test-executions.js:1510]: Should duration be null here?

**Review by Jkovarik — CHANGES_REQUESTED** [review]
@jennyhliu  thanks for addressing the changes, just one more comment re: that unit and a comment/question. 

Thanks for your patience re: me getting back to you on this.

File: packages/api/tests/endpoints/test-executions.js (hunk 8 above)
Comment (jennyhliu) [reply packages/api/tests/endpoints/test-executions.js]: [d8bfba7](https://github.com/nasa/cumulus/pull/3186/commits/d8bfba7312a9909a872eda1be3158cbfc601a3e5)

File: packages/db/src/translate/executions.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jennyhliu) [reply packages/db/src/translate/executions.ts]: [d8bfba7](https://github.com/nasa/cumulus/pull/3186/commits/d8bfba7312a9909a872eda1be3158cbfc601a3e5)

**Review by Jkovarik — APPROVED** [review]
Looks good!  Nice work!

**Pull request merged**